core/src/translate/param.rs

```diff
@@ -0,0 +1,529 @@
+use {
+    super::TranslateError,
+    crate::{
+        ast::{AstLiteral, DataType, DateTimeField, Expr},
+        data::{Interval, Point},
+    },
+    bigdecimal::BigDecimal,
+    chrono::{NaiveDate, NaiveDateTime, NaiveTime, Timelike},
+    rust_decimal::Decimal,
+    std::net::IpAddr,
+    uuid::Uuid,
+};
+
+#[derive(Debug, Clone)]
+pub enum ParamLiteral {
+    Literal(AstLiteral),
+    TypedString {
+        data_type: DataType,
+        value: String,
+    },
+    Interval {
+        expr: Box<ParamLiteral>,
+        leading_field: Option<DateTimeField>,
+        last_field: Option<DateTimeField>,
+    },
+}
+
+impl ParamLiteral {
+    pub const fn null() -> Self {
+        Self::Literal(AstLiteral::Null)
+    }
+
+    pub fn into_expr(self) -> Expr {
+        match self {
+            ParamLiteral::Literal(literal) => Expr::Literal(literal),
+            ParamLiteral::TypedString { data_type, value } => {
+                Expr::TypedString { data_type, value }
+            }
+            ParamLiteral::Interval {
+                expr,
+                leading_field,
+                last_field,
+            } => Expr::Interval {
+                expr: Box::new(expr.into_expr()),
+                leading_field,
+                last_field,
+            },
+        }
+    }
+}
+
+fn into_number_literal(value: impl ToString) -> Result<AstLiteral, TranslateError> {
```

Comment:
Would you consider taking this argument by reference instead? It’s only read (not consumed) in the body, so borrowing it would avoid unnecessary moves.

```rust
fn into_number_literal(value: &impl ToString)
```